File: src/camera.ts
import { mat4, vec3 } from "gl-matrix";

export class Camera {
  position: vec3;
  yaw: number; // radians – rotation around Y axis
  pitch: number; // radians – rotation around X axis

  private readonly sensitivity = 0.002;
  private readonly speed = 8.0; // units per second

  constructor() {
    this.position = vec3.fromValues(32, 20, 32);
    this.yaw = 0;
    this.pitch = -0.3;
  }

  /** Returns the unit forward vector derived from yaw + pitch. */
  getForward(): vec3 {
    return vec3.fromValues(
      Math.cos(this.pitch) * Math.sin(this.yaw),
      Math.sin(this.pitch),
      -Math.cos(this.pitch) * Math.cos(this.yaw),
    );
  }

  /** Returns the unit right vector (always horizontal). */
  getRight(): vec3 {
    return vec3.fromValues(
      Math.cos(this.yaw),
      0,
      Math.sin(this.yaw),
    );
  }

  /** Forward projected onto the XZ plane – used for walking so pitch doesn't affect movement. */
  getFlatForward(): vec3 {
    return vec3.fromValues(
      Math.sin(this.yaw),
      0,
      -Math.cos(this.yaw),
    );
  }

  rotate(dx: number, dy: number): void {
    this.yaw += dx * this.sensitivity;
    this.pitch -= dy * this.sensitivity;
    // Clamp pitch to avoid flipping
    const limit = (89 * Math.PI) / 180;
    this.pitch = Math.max(
      -limit,
      Math.min(limit, this.pitch),
    );
  }

  moveForward(dt: number): void {
    const fwd = this.getForward();
    vec3.scaleAndAdd(
      this.position,
      this.position,
      fwd,
      this.speed * dt,
    );
  }

  moveRight(dt: number): void {
    const right = this.getRight();
    vec3.scaleAndAdd(
      this.position,
      this.position,
      right,
      this.speed * dt,
    );
  }

  moveUp(dt: number): void {
    this.position[1] += this.speed * dt;
  }

  getViewMatrix(): mat4 {
    const target = vec3.create();
    vec3.add(target, this.position, this.getForward());
    const view = mat4.create();
    mat4.lookAt(
      view,
      this.position,
      target,
      vec3.fromValues(0, 1, 0),
    );
    return view;
  }

  getProjectionMatrix(aspect: number): mat4 {
    const proj = mat4.create();
    mat4.perspective(
      proj,
      (70 * Math.PI) / 180,
      aspect,
      0.1,
      500.0,
    );
    return proj;
  }

  /**
   * Same field-of-view as getProjectionMatrix() but uses the zero-to-one
   * depth range expected by WebGPU (clip z ∈ [0, 1] rather than [-1, 1]).
   * Use this matrix when uploading to the WebGPU renderer's UBOs.
   */
  getProjectionMatrixZO(aspect: number): mat4 {
    const proj = mat4.create();
    mat4.perspectiveZO(
      proj,
      (70 * Math.PI) / 180,
      aspect,
      0.1,
      500.0,
    );
    return proj;
  }

  setPose(
    position: [number, number, number],
    yaw: number,
    pitch: number,
  ): void {
    this.position[0] = position[0];
    this.position[1] = position[1];
    this.position[2] = position[2];
    this.yaw = yaw;
    this.pitch = pitch;
  }

  getPose(): {
    position: [number, number, number];
    yaw: number;
    pitch: number;
  } {
    return {
      position: [
        this.position[0],
        this.position[1],
        this.position[2],
      ],
      yaw: this.yaw,
      pitch: this.pitch,
    };
  }
}

File: game/src/engine/camera.rs
use glam::{Mat4, Vec3};

const SENSITIVITY: f32 = 0.002;

pub struct Camera {
    pub position: Vec3,
    /// Yaw in radians (rotation around Y axis).
    pub yaw: f32,
    /// Pitch in radians (rotation around X axis).
    pub pitch: f32,
}

impl Camera {
    pub fn new() -> Self {
        Self {
            position: Vec3::new(32.0, 20.0, 32.0),
            yaw: 0.0,
            pitch: -0.3,
        }
    }

    /// Unit forward vector from yaw + pitch.
    pub fn forward(&self) -> Vec3 {
        Vec3::new(
            self.pitch.cos() * self.yaw.sin(),
            self.pitch.sin(),
            -self.pitch.cos() * self.yaw.cos(),
        )
    }

    /// Unit right vector (always horizontal).
    pub fn right(&self) -> Vec3 {
        Vec3::new(self.yaw.cos(), 0.0, self.yaw.sin())
    }

    /// Forward projected onto XZ — used for walking so
    /// pitch doesn't affect movement direction.
    pub fn flat_forward(&self) -> Vec3 {
        Vec3::new(
            self.yaw.sin(),
            0.0,
            -self.yaw.cos(),
        )
    }

    pub fn rotate(&mut self, dx: f32, dy: f32) {
        self.yaw += dx * SENSITIVITY;
        self.pitch -= dy * SENSITIVITY;
        let limit = 89.0_f32.to_radians();
        self.pitch = self.pitch.clamp(-limit, limit);
    }

    /// View matrix (right-handed, Y-up).
    pub fn view_matrix(&self) -> Mat4 {
        Mat4::look_to_rh(
            self.position,
            self.forward(),
            Vec3::Y,
        )
    }

    /// Perspective projection matrix.
    pub fn projection_matrix(
        &self,
        aspect: f32,
    ) -> Mat4 {
        Mat4::perspective_rh(
            70.0_f32.to_radians(),
            aspect,
            0.1,
            1000.0,
        )
    }

    pub fn get_pose(&self) -> CameraPose {
        CameraPose {
            position: self.position.to_array(),
            yaw: self.yaw,
            pitch: self.pitch,
        }
    }

    pub fn set_pose(&mut self, pose: &CameraPose) {
        self.position = Vec3::from(pose.position);
        self.yaw = pose.yaw;
        self.pitch = pose.pitch;
    }
}

pub struct CameraPose {
    pub position: [f32; 3],
    pub yaw: f32,
    pub pitch: f32,
}
